File: Week3/ArrayProblem/FindDuplicateNumber.java
import java.util.HashMap;

public class FindDuplicateNumber {

/*
 * Given an array of integers nums containing n + 1 integers where each integer is in the range [1, n] inclusive.
 * There is only one repeated number in nums, return this repeated number.
 * You must solve the problem without modifying the array nums and uses only constant extra space.
 * T(c) -> O(n^2), S(c) -> O(1)
 */

    static int findDuplicate(int[] nums) {
        int n = nums.length;
        // check frequency of each number
        for(int i=0;i<n;i++)
        {
            int count = 1;
            for(int j=i+1;j<n;j++)
            {
                if(nums[i] == nums[j])
                {
                    count++;
                }
            }
            
            // if frequency is greater than 1, means duplicate
            if(count > 1)
            {
                return nums[i];
            }
        }
        // by default case, duplicate not found
        return -1;
    }

    /*
     * Using hashing
     * T(c) -> O(n), S(c) -> O(n)
    */
    static int findDuplicateBetter(int[] nums) {
        int n = nums.length;
        
        HashMap<Integer,Integer> map = new HashMap<>();
        
        // check frequency of each number
        for(int i=0;i<n;i++)
        {
            map.put(nums[i], map.getOrDefault(nums[i], 0)+1);
            
            // if frequency is greater than 1, duplicate
            if(map.get(nums[i]) > 1)
            {
                return nums[i];
            }
        }
         
        // by default case, duplicate not found
        return -1;
    }
    
    /*
     * Using floyd's tortoise and hare algo (cycle detection)
     * T(c) -> O(n), S(c) -> O(1)
     */
    static int findDuplicateOptimal(int[] nums) {
        int slow = nums[0];
        int fast = nums[0];
        
        // move fast pointer 2 step and slow pointer 1 step until both meets at one point
        do{
            slow = nums[slow];
            fast = nums[nums[fast]];
        }
        while(slow != fast);
        
        /* point fast pointer in the beginning and start to travarse again 
         this  time both slow and fast pointer will go 1 step until both will meet again */
        fast = nums[0];
        while(slow != fast)
        {
            fast = nums[fast];
            slow = nums[slow];
        }
        return fast;
    }


    public static void main(String[] args) {
        int arr[] = {1,3,4,2,2};
        System.out.println("Duplicate = "+findDuplicate(arr));
        System.out.println("Duplicate = "+findDuplicateBetter(arr));
        System.out.println("Duplicate = "+findDuplicateOptimal(arr));
    }

}

File: Week3/Python/ArrayProblems/FindDuplicateNumber.py
"""
    Given an array of integers nums containing n + 1 integers where each integer is in the range [1, n] inclusive.
    There is only one repeated number in nums, return this repeated number.
    You must solve the problem without modifying the array nums and using only constant extra space.

    https://leetcode.com/problems/find-the-duplicate-number/
"""

def findDuplicate(nums):
    """ T(c) -> O(n), S(c) -> O(n)"""

    freq = {}
    
    # calculate frequency of elements
    for i in range(len(nums)):
        elem = nums[i]
        if freq.get(elem):
            freq[elem] += 1
        else:
            freq[elem] = 1
            
    repeated = -1
    
    # check of repeated number
    for key, value in freq.items():
        if value > 1:
            repeated = key
    
    return repeated


def findDuplicateEfficient(nums):
    """ T(c) -> O(n), S(c) -> O(1)"""

    # using floyd's cycle detection or tortoise and hare algo
    # slow and fast pointer will point to the first element
    slow = nums[0]
    fast = nums[0]
    
    # slow will move one step and fast will move two step until both meet again
    while True:
        slow = nums[slow]
        fast = nums[nums[fast]]
        if slow == fast:
            break
            
    # both met at a point, so cycle exist, reset fast pointerto start of array
    fast = nums[0]
    # this time move fast and slow both one step until it meet again
    while fast != slow:
        fast = nums[fast]
        slow = nums[slow]
    
    # that point where both meet again, is the duplicate number
    return slow


arr = [3,1,3,4,3,2]
print(arr)
print(findDuplicate(arr))
print(findDuplicateEfficient(arr))
